Core/ContentBlockerRulesUserScript.swift

```diff
@@ -71,12 +72,54 @@ public class ContentBlockerRulesUserScript: NSObject, UserScript {
         }
     }
     
-    private func trackerFromUrl(_ urlString: String, blocked: Bool) -> DetectedTracker? {
-        let knownTracker = TrackerDataManager.shared.findTracker(forUrl: urlString)
-        if let entity = TrackerDataManager.shared.findEntity(byName: knownTracker?.owner?.name ?? "") {
-            return DetectedTracker(url: urlString, knownTracker: knownTracker, entity: entity, blocked: blocked)
+    private func trackerFromUrl(_ trackerUrlString: String, pageUrlString: String, blockable: Bool) -> DetectedTracker? {
+        guard let knownTracker = TrackerDataManager.shared.findTracker(forUrl: trackerUrlString) else {
+            return nil
+        }
+
+        let blocked: Bool
+
+        if knownTracker.hasExemption(for: trackerUrlString, pageUrlString: pageUrlString) {
+            blocked = false
+        } else {
+            blocked = blockable
+        }
+
+        if let entity = TrackerDataManager.shared.findEntity(byName: knownTracker.owner?.name ?? "") {
+            return DetectedTracker(url: trackerUrlString, knownTracker: knownTracker, entity: entity, blocked: blocked)
         }
         
         return nil
     }
 }
+
+fileprivate extension KnownTracker {
+
+    func hasExemption(for trackerUrlString: String, pageUrlString: String) -> Bool {
+        let range = NSRange(location: 0, length: trackerUrlString.utf16.count)
+
+        // For each rule, where the rule has an `ignore` action:
+        //
+        // 1. Create a regular expression from the rule
+        // 2. Check if the rule matches the tracker URL
+        // 3. If it matches, check if the rule contains an exemption for the page's host URL
+        //    3a. If exempt, continue checking rules
+        //    3b. If not exempt, return true – an exemption was found, no further checking needed
+
+        for rule in rules ?? [] where rule.action == .ignore {
+            guard let pattern = rule.rule,
+                  let regex = try? NSRegularExpression(pattern: pattern, options: []) else { continue }
+
+            if regex.firstMatch(in: trackerUrlString, options: [], range: range) != nil {
+                if let host = URL(string: pageUrlString)?.host, rule.exceptions?.domains?.contains(host) == true {
```

Comment:
Can move this check up into the `guard` statement to avoid matching a regex only to find an exception